### PR TITLE
Run C++ tests on CI with run_test.py

### PR DESCRIPTION
**Pull request by huydhn**
After https://github.com/pytorch/pytorch/pull/99559, we can now run C++ test with `run_test.py`.  Although advance features such as `--import-slow-tests` and `--import-disabled-tests` won't work for now, there will still be a gain in reliability and performance as C++ can now be retried and run in parallel.

This covers all C++ tests in the CI including aten, libtorch, and Vulkan C++ tests across all platforms Linux, Windows, MacOS.

Notes:
* To support C++ test discovery, the env variable `CPP_TESTS_DIR` can be set to where the C++ test binaries is located
* Support pytest -k argument via run_test as this is used by pytest-cpp to replace `--gtest-filter`
* The XML output is in pytest format, but it's ok now because we don't have slow test or flaky test support for C++ test yet
* ~~I need to figure out why conftest.py doesn't work when I invoke pytest directly for C++ test, so `--sc` is not available for C++ tests at the moment.  Proper pytest plugin like stepwise works fine though.  I'll investigate and fix it in a separate PR~~ Found the cause, `conftest.py` is per directory and needs to be in any arbitrary directory that holds C++ test
* Two tests `test_api` and `test_tensorexpr` timed out on ASAN, I suspect that ASAN is now used on top of the python executable, which is slower than running native C++ code.  IMO, it's ok to run these tests as before on ASAN for now

### CODE REVIEW

**Review by clee2000 — APPROVED** [review]

File: .ci/pytorch/test.sh [hunk 1]
```diff
@@ -547,43 +547,60 @@ test_libtorch() {
...
-      LTC_TS_CUDA=1 "$TORCH_BIN_DIR"/test_lazy  --gtest_output=xml:$TEST_REPORTS_DIR/test_lazy.xml
+      LTC_TS_CUDA=1 python test/run_test.py --cpp --verbose -i cpp/test_lazy
     else
-      "$TORCH_BIN_DIR"/test_lazy  --gtest_output=xml:$TEST_REPORTS_DIR/test_lazy.xml
+      python test/run_test.py --cpp --verbose -i cpp/test_lazy
     fi
 
-    python test/cpp/jit/tests_setup.py shutdown
+    # Cleaning up test artifacts in the test folder
+    pushd test
```
Comment (ZainRizvi) [inline .ci/pytorch/test.sh:578]: I'm curious why the `pushd/popd` was needed now when it wasn't needed before

File: aten/tools/run_tests.sh [hunk 2]
```diff
@@ -3,75 +3,78 @@ set -x
...
-  ./cuda_tensor_interop_test
+  python test/run_test.py --cpp --verbose -i cpp/cuda_tensor_interop_test
 fi
 if [[ -x ./cuda_complex_test ]]; then
-  ./cuda_complex_test
+  python test/run_test.py --cpp --verbose -i cpp/cuda_complex_test
 fi
 if [[ -x ./cuda_complex_math_test ]]; then
-  ./cuda_complex_math_test
+  python test/run_test.py --cpp --verbose -i cpp/cuda_complex_math_test
```
Comment (ZainRizvi) [inline aten/tools/run_tests.sh:64]: Why make these verbose? Does this turn them into "as verbose as normal python tests" mode, or do cpp tests need to be extra verbose for some reason?

**Review by ZainRizvi — APPROVED** [review]

File: .ci/pytorch/test.sh (hunk 1 above)
Comment (huydhn) [reply .ci/pytorch/test.sh:578]: This script `test/cpp/jit/tests_setup.py` creates and cleans up a dummy JIT output for testing to the current directory.  Previously, when the C++ tests was run directly as an executable, there was no need to change the directory. On the other hand, `run_test.py` changes the working directory to `test`, thus I need to do the same here to create that dummy output in `test` folder instead.

**Pull request closed without merging**